File: frontend/src/domain/riskScore.ts
export function roundPositiveRatioHalfUp(
  numerator: number,
  denominator: number,
): number {
  const scale = 10_000n;
  const divisor = BigInt(denominator);
  const scaledNumerator = BigInt(numerator) * scale;
  let rounded = scaledNumerator / divisor;
  if ((scaledNumerator % divisor) * 2n >= divisor) rounded += 1n;
  return Number(rounded) / Number(scale);
}

File: backend/risk_score.py
def round_positive_ratio_half_up(
    numerator: int,
    denominator: int,
    decimal_places: int = 4,
) -> float:
    scale = 10**decimal_places
    quotient, remainder = divmod(numerator * scale, denominator)
    if remainder * 2 >= denominator:
        quotient += 1
    return quotient / scale
